### PR TITLE
predicate rng ops

### LINKED ISSUE #1084: low performance for kernel with padded dim and RNG Ops
In this case, bdimx is padded from 106 to 128 and the prediate in else branch for RNG Ops is removed. This lead to performance regression compared with unpadded version as discussed in https://github.com/NVIDIA/Fuser/pull/988. Performance data for different tests:

Test | latency, ms
-- | --
padded (main branch) | 0.74
unpadded | 0.56
padded with predicate | 0.58
unpadded with predicate | 0.56

To solve this issue, we can:
(1) don't eliminate predicate for RNG Ops
(2) use unpadded kernel.

Script to reproduce:

```
import math
import torch
from nvfuser import FusionDefinition, DataType
from nvfuser.pytorch_utils import torch_dtype_to_nvfuser_dtype


def nvfuser_scale_mask_softmax_dropout(fd: FusionDefinition, shape, input_dtype, prob=0.20) -> None:
    n_head = shape[1]
    embed_dim_per_head = 64
    batch_size = shape[0]
    seq_len = shape[-1]
    r_sqrt_nhead = 1.0 / math.sqrt(embed_dim_per_head)

    # batch size, head count, sequence length, sequence length
    T0 = fd.define_tensor(shape=[-1, -1, -1, -1], contiguity=[True, True, True, True], dtype=input_dtype, is_cpu=False)
    T1 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=input_dtype, is_cpu=False)
    if input_dtype == DataType.Half:
        T0 = fd.ops.cast(T0, dtype=DataType.Float)
        T1 = fd.ops.cast(T1, dtype=DataType.Float)
    T1 = fd.ops.broadcast_in_dim(T1, shape=shape, broadcast_dims=[2, 3])
    S2 = fd.define_scalar(r_sqrt_nhead, dtype=DataType.Double)
    T3 = fd.ops.mul(T0, S2)
    T4 = fd.ops.slice(T1, start_indices=[0, 0, 0, 0], end_indices=[1, 1, seq_len, seq_len], strides=[1, 1, 1, 1])
    S5 = fd.define_scalar(0.00000, dtype=DataType.Double)
    T6 = fd.ops.eq(S5, T4)
    T7 = fd.ops.broadcast_in_dim(T6, shape=[batch_size, n_head, seq_len, seq_len], broadcast_dims=[0, 1, 2, 3])
    S8 = fd.define_scalar(float("-inf"), dtype=DataType.Double)
    T9 = fd.ops.where(T7, S8, T3)
    T13 = fd.ops.max(T9, axes=[3], keepdim=True, dtype=DataType.Null)
    T16 = fd.ops.sub(T9, T13)
    T17 = fd.ops.exp(T16)
    T21 = fd.ops.sum(T17, axes=[3], keepdim=True, dtype=DataType.Null)
    r_T21 = fd.ops.reciprocal(T21)
    T24 = fd.ops.mul(T17, r_T21)
    S25 = fd.define_scalar(batch_size, dtype=DataType.Int)
    S26 = fd.define_scalar(n_head, dtype=DataType.Int)
    S27 = fd.define_scalar(seq_len, dtype=DataType.Int)
    S28 = fd.define_scalar(seq_len, dtype=DataType.Int)
    S29 = fd.define_scalar(0.00000, dtype=DataType.Double)
    S30 = fd.define_scalar(1.00000, dtype=DataType.Double)
    T31 = fd.ops.uniform(S29, S30, shape=[S25, S26, S27, S28], dtype=DataType.Float)
    S32 = fd.define_scalar(1.0 - prob, dtype=DataType.Double)
    T33 = fd.ops.lt(T31, S32)
    T34 = fd.ops.cast(T33, dtype=DataType.Float)
    T35 = fd.ops.mul(T24, T34)
    S36 = fd.define_scalar(1.0 / (1.0 - prob), dtype=DataType.Double)
    T37 = fd.ops.mul(T35, S36)

    if input_dtype == DataType.Half:
        T37 = fd.ops.cast(T37, dtype=DataType.Half)

    fd.add_output(T37)


if __name__ == "__main__":
    torch_dtype = torch.half
    batch = 2
    seq_len = 4224
    n_head = 2
    inputs = [
        torch.randn(batch, n_head, seq_len, seq_len, dtype=torch_dtype, device="cuda"),
        torch.randn(seq_len, seq_len, dtype=torch_dtype, device="cuda")
    ]

    with FusionDefinition() as fd:
        nvfuser_scale_mask_softmax_dropout(fd, inputs[0].shape, torch_dtype_to_nvfuser_dtype(torch_dtype))

    nvfuser_output = fd.execute(inputs)

```

### PR DESCRIPTION
**Pull request by liqiangxl**
Fix #1084. After this PR, predicate for RNG Ops will NOT be eliminated, e.g.
```
        if (b63) {
          nvfuser_index_t linear_index2259 = (i58 + (i64 + nvfuser_zero));
          nvfuser_index_t rng_subseq2259 = linear_index2259 / 4;
          nvfuser_index_t rng_component2259 = linear_index2259 % 4;
          nvfuser_index_t rng_offset2259 = (((ptr2 == nullptr) ? i3 : ((*ptr2) + i3)) / 4LL);
          if (rng_subseq != rng_subseq2259 || rng_offset != rng_offset2259) {
            rng_result = philox(((ptr0 == nullptr) ? i1 : (*ptr0)), rng_subseq2259, rng_offset2259);
            rng_subseq = rng_subseq2259;
            rng_offset = rng_offset2259;
          }
          T20[0] = rng_uniform_rangef(rng_result, rng_component2259, 0.00000000000000000e+00, 1.00000000000000000e+00);
        }
```

The original case was tested on H100,
(1) main branch, predicate for RNG Ops eliminated: kernel1 run in 0.476512 ms, achieved: 673.979 GB/s
(2) this PR, predicate for RNG Ops not eliminated: kernel1 run in 0.3376 ms, achieved: 951.301 GB/s

Introduced 4 additional benchmark scenarios within `NvFuserScheduler_Softmax_Dropout_Inner_fp16`, resulting in bandwidth enhancements ranging from 10% to 20%.
image

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
Code diff in cases with RNG Ops is expected.

**Comment by naoyam** [comment]
Have you looked at performance changes of other fusions with rng?

**Comment by liqiangxl** [comment]
> Have you looked at performance changes of other fusions with rng?

Yes. The other one is `NvFuserScheduler_BiasDropoutAddLayernorm*` in `bert.cpp`, no change in performance since no padding.

**Comment by liqiangxl** [comment]
> This could change the performance even without padding, right?

For unpaded case, only if there are non-divisible split, e.g. in the [original case](https://github.com/NVIDIA/Fuser/pull/988#:~:text=diverged%20rng_result%3A%20tid%3D104%20i55%3D4%0Adiverged%20rng_result%3A%20tid%3D105%20i55%3D4) it only inflence the 2 threads when processing the last persistent batch.
Otherwise, no threads goes to the else branch.
results for `NvFuserScheduler_BiasDropoutAddLayernorm`
image

**Comment by naoyam** [comment]
Yes, it depends on the ratio of threads taking the else path, and that depends on problem sizes and our heuristics. My concern is if we were only looking at "easy" problem sizes, we shouldn't see any difference, but that may not be representative and also not covered by our current C++ benchmarks.

**Comment by liqiangxl** [comment]
> Yes, it depends on the ratio of threads taking the else path, and that depends on problem sizes and our heuristics. My concern is if we were only looking at "easy" problem sizes, we shouldn't see any difference, but that may not be representative and also not covered by our current C++ benchmarks.

You are right. our current C++ benchmarks do not cover the cases where both RNG Ops and paddings are used. So I added 4 additional benchmark scenarios within `NvFuserScheduler_Softmax_Dropout_Inner_fp16`.

**Comment by naoyam** [comment]
Aren't they padded benchmarks?

**Comment by liqiangxl** [comment]
> Aren't they padded benchmarks?

Yes. The newly added cases are padded. We want them to be padded so we have some threads go to the else branch.

**Comment by naoyam** [comment]
Can you add a test (not benchmark) that has rng and show a performance difference even without padding? This should be possible, right?

Also, have you measured the performances of V100 and A100? Curious if this independent scheduling has different effects on older devices.

**Comment by liqiangxl** [comment]
> Can you add a test (not benchmark) that has rng and show a performance difference even without padding? This should be possible, right?
> 
> Also, have you measured the performances of V100 and A100? Curious if this independent scheduling has different effects on older devices.

(1)Yes. I'll add a test has rng but without padding.
(2) The CPP benchmark `NvFuserScheduler_Softmax_Dropout_Inner_fp16` was measured on A100 and `NvFuserScheduler_BiasDropoutAddLayernorm` was measured on H100. Device was picked based on which is available. 
(3)The original case was measured on both H100 and A100.
```
The original case was tested on H100,
(1) main branch, predicate for RNG Ops eliminated: kernel1 run in 0.476512 ms, achieved: 673.979 GB/s
(2) this PR, predicate for RNG Ops not eliminated: kernel1 run in 0.3376 ms, achieved: 951.301 GB/s
```

Results on A100:
Test | latency, ms
-- | --
padded (main branch) | 0.74
unpadded | 0.56
padded with predicate | 0.58
unpadded with predicate | 0.56

**Comment by naoyam** [comment]
Everything looks good. Just let me know when you add the test.

**Comment by liqiangxl** [comment]
> Everything looks good. Just let me know when you add the test.

Test is added.

File: test/test_gpu3.cpp [hunk 1]
```diff
@@ -9662,18 +9662,29 @@ TEST_F(NVFuserTest, SoftmaxDropout) {
...
-      rng_consumers.insert(tv);
+  // check RNGOp is predicated without else branch.
+  class PredicateChecker : public kir::IrVisitor {
+   public:
+    using kir::IrVisitor::handle;
+    bool predicate_rngop = false;
+
+   private:
+    void handle(RNGOp* uop) final {
+      for (auto expr : scope_exprs_) {
```
Comment (naoyam) [inline test/test_gpu3.cpp:9673]: Shouldn't we only check the innermost `IfThenElse`?

**Comment by naoyam** [comment]
>(1)Yes. I'll add a test has rng but without padding.

Where is this? For the added test, the comment says it's padded

**Comment by liqiangxl** [comment]
> > (1)Yes. I'll add a test has rng but without padding.
> 
> Where is this? For the added test, the comment says it's padded

Wrong case! Let me change the test.

File: test/test_gpu3.cpp (hunk 1 above)
Comment (liqiangxl) [reply test/test_gpu3.cpp:9676]: I think the key is to find the one in the `else` branch. So here I used `expr->as<kir::IfThenElse>()->hasElse()`. It captures error if predicate is eliminated.

**Comment by liqiangxl** [comment]
> > (1)Yes. I'll add a test has rng but without padding.
> 
> Where is this? For the added test, the comment says it's padded

For unpadded case, the difference is neglectable.

**Comment by naoyam** [comment]
> > > (1)Yes. I'll add a test has rng but without padding.
> > 
> > 
> > Where is this? For the added test, the comment says it's padded
> 
> For unpadded case, the difference is neglectable.

Could you create a synthetic fusion and manually schedule to expose the original issue without padding?

**Review by naoyam — APPROVED** [review]
LGTM
